libsawtooth/src/journal/state_verifier.rs

```diff
@@ -0,0 +1,337 @@
+
+    let start_block = match start_block {
+        Some(block) => block,
+        None => {
+            info!("Skipping state verification: chain head's state root is present");
+            return Ok(());
+        }
+    };
+
+    let previous_state_root = prev_state_root.unwrap_or_else(|| initial_state_root.to_string());
+
+    info!(
+        "Recomputing missing state from block {}",
+        start_block.header().block_num()
+    );
+
+    let blocks = commit_store.get_block_by_height_iter(
+        Some(start_block.header().block_num()),
+        ByHeightDirection::Increasing,
+    );
+
+    process_blocks(
+        &previous_state_root,
+        blocks,
+        execution_task_submitter,
+        merkle_state,
+        state_view_factory,
+        scheduler_factory,
+    )
+}
+
+/// Search through the blockstore and return a tuple containing:
+///   - the first block with a missing state root
+///   - the state root of that blocks predecessor
+fn search_for_present_state_root(
+    commit_store: &CommitStore,
+    state_view_factory: &StateViewFactory,
+    initial_state_root: &str,
+) -> Result<(Option<BlockPair>, Option<String>), StateVerificationError> {
+    // If there is no chain to process, then we are done.
+    let block = match commit_store.get_chain_head() {
+        Ok(block) => block,
+        Err(DatabaseError::NotFoundError(_)) => return Ok((None, None)),
+        Err(err) => {
+            return Err(StateVerificationError::InvalidChainError(format!(
+                "Unable to check for chain head: {}",
+                err
+            )))
+        }
+    };
+
+    // Check the head first
+    if state_view_factory
+        .create_view::<SettingsView>(block.header().state_root_hash())
+        .is_ok()
+    {
+        return Ok((None, None));
+    }
+
+    let mut previous_state_root = initial_state_root.to_string();
+    for block in commit_store.get_block_by_height_iter(None, ByHeightDirection::Increasing) {
+        if state_view_factory
+            .create_view::<SettingsView>(block.header().state_root_hash())
+            .is_err()
+        {
+            return Ok((Some(block), Some(previous_state_root)));
+        }
+        previous_state_root = hex::encode(block.header().state_root_hash());
+    }
+
+    // This should never happen, since we already checked that the chain head
+    // didn't have a state root
+    Err(StateVerificationError::InvalidChainError(
+        "Chain head state both missing but all blocks had state root present".to_string(),
+    ))
+}
+
+/// Interate through commited blocks and build the block's state if it is missing
+fn process_blocks(
+    initial_state_root: &str,
+    blocks: CommitStoreByHeightIterator,
+    execution_task_submitter: ExecutionTaskSubmitter,
+    merkle_state: CborMerkleState,
+    state_view_factory: &StateViewFactory,
+    scheduler_factory: Box<dyn SchedulerFactory>,
```

Comment:
I think all of these can be taken by reference? Would that prevent cloning them where this gets called in the validator?

Reply:
updated